### ISSUE TITLE
ICE with uninhabited and dynamically sized types

### ISSUE THREAD

**Issue by svenknobloch** [rust-lang/rust#82954]
### Code

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8665e9b4c18533ba0d52f61841328baf)

```Rust
use std::marker::PhantomData;

pub enum Empty {}

struct Test<A, B: ?Sized> {
    phantom: PhantomData<(A, B)>,
}

fn main() {
    let x: Test<Empty, [usize]>;
}
```

### Error output

```
   Compiling playground v0.0.1 (/playground)
warning: unused variable: `x`
  --> src/main.rs:10:9
   |
10 |     let x: Test<Empty, [usize]>;
   |         ^ help: if this is intentional, prefix it with an underscore: `_x`
   |
   = note: `#[warn(unused_variables)]` on by default

thread 'rustc' panicked at 'assertion failed: layout.abi.is_uninhabited()', compiler/rustc_middle/src/ty/layout.rs:205:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.50.0 (cb75ad5db 2021-02-10) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [layout_raw] computing layout of `(Empty, [usize])`
end of query stack
warning: 1 warning emitted

error: could not compile `playground`

To learn more, run the command again with --verbose.
```

The error seems to only occur when using an uninhabited type with a dynamically sized one in a tuple.
Inhabited types with dynamically sized and uninhabited types with sized types both work.

A simple workaround for now is to box the DST in the PhantomData.

**Comment by SNCPlay42** [comment]
This compiles successfully on 1.32 [(godbolt)](https://rust.godbolt.org/z/hhqvz9). It appears to require `-C debuginfo=2` to reproduce.

@rustbot label regression-from-stable-to-stable

**Comment by jyn514** [comment]
Slightly smaller:
```rust
use std::marker::PhantomData;

pub enum Empty {}

fn main() {
    let _x: PhantomData<(Empty, [usize])>;
}
```

**Comment by hameerabbasi** [comment]
Bisected:

```
found 5 bors merge commits in the specified range
  commit[0] 2018-12-20UTC: Auto merge of #56845 - GuillaumeGomez:const-docs, r=oli-obk
  commit[1] 2018-12-20UTC: Auto merge of #54125 - varkor:less-conservative-uninhabitedness-check, r=nikomatsakis
  commit[2] 2018-12-21UTC: Auto merge of #55798 - GuillaumeGomez:version-display-associated-const, r=QuietMisdreavus
  commit[3] 2018-12-21UTC: Auto merge of #56813 - oli-obk:main_🧶, r=pnkfelix
  commit[4] 2018-12-21UTC: Auto merge of #56779 - adrian-budau:master, r=alexcrichton
ERROR: no commits between 09d6ab90e556bf692ff3f8790d97b3ca4fee94b0 and e40548bc43f2a0375a466c98e174e71561dc98d2 within last 167 days
```

**Comment by apiraino** [comment]
Assigning `P-medium` as discussed as part of the [Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2382954.20ICE.20with.20uninhabited.20and.20dynamically.20sized.20types) and removing `I-prioritize`.

@rustbot label -I-prioritize +P-medium

**Comment by jyn514** [comment]
> Bisected:
> 
> ```
> found 5 bors merge commits in the specified range
>   commit[0] 2018-12-20UTC: Auto merge of #56845 - GuillaumeGomez:const-docs, r=oli-obk
>   commit[1] 2018-12-20UTC: Auto merge of #54125 - varkor:less-conservative-uninhabitedness-check, r=nikomatsakis
>   commit[2] 2018-12-21UTC: Auto merge of #55798 - GuillaumeGomez:version-display-associated-const, r=QuietMisdreavus
>   commit[3] 2018-12-21UTC: Auto merge of #56813 - oli-obk:main_🧶, r=pnkfelix
>   commit[4] 2018-12-21UTC: Auto merge of #56779 - adrian-budau:master, r=alexcrichton
> ERROR: no commits between 09d6ab90e556bf692ff3f8790d97b3ca4fee94b0 and e40548bc43f2a0375a466c98e174e71561dc98d2 within last 167 days
> ```

https://github.com/rust-lang/rust/pull/54125 looks like it's the relevant PR, but it's not actually a bug, it just extends the checks to more items. I don't think this is really a regression, the bug was always there (only the ICE is a regression).

**Comment by tmiasko** [comment]
A struct layout is uninhabited only if it *sized* and has any uninhabited fields (since #50622). This creates discrepancy with type-level view of things.

https://github.com/rust-lang/rust/blob/61edfd591cedff66fca639c02f66984f6271e5a6/compiler/rustc_middle/src/ty/layout.rs#L481-L483

**Comment by Alexendoo** [comment]
Fixed by #90854
